(self.webpackChunk=self.webpackChunk||[]).push([[367],{3905:(e,t,n)=>{"use strict";n.d(t,{Zo:()=>d,kt:()=>h});var a=n(7294);function r(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function i(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);t&&(a=a.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,a)}return n}function o(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?i(Object(n),!0).forEach((function(t){r(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):i(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function l(e,t){if(null==e)return{};var n,a,r=function(e,t){if(null==e)return{};var n,a,r={},i=Object.keys(e);for(a=0;a<i.length;a++)n=i[a],t.indexOf(n)>=0||(r[n]=e[n]);return r}(e,t);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);for(a=0;a<i.length;a++)n=i[a],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(r[n]=e[n])}return r}var s=a.createContext({}),p=function(e){var t=a.useContext(s),n=t;return e&&(n="function"==typeof e?e(t):o(o({},t),e)),n},d=function(e){var t=p(e.components);return a.createElement(s.Provider,{value:t},e.children)},c={inlineCode:"code",wrapper:function(e){var t=e.children;return a.createElement(a.Fragment,{},t)}},u=a.forwardRef((function(e,t){var n=e.components,r=e.mdxType,i=e.originalType,s=e.parentName,d=l(e,["components","mdxType","originalType","parentName"]),u=p(n),h=r,m=u["".concat(s,".").concat(h)]||u[h]||c[h]||i;return n?a.createElement(m,o(o({ref:t},d),{},{components:n})):a.createElement(m,o({ref:t},d))}));function h(e,t){var n=arguments,r=t&&t.mdxType;if("string"==typeof e||r){var i=n.length,o=new Array(i);o[0]=u;var l={};for(var s in t)hasOwnProperty.call(t,s)&&(l[s]=t[s]);l.originalType=e,l.mdxType="string"==typeof e?e:r,o[1]=l;for(var p=2;p<i;p++)o[p]=n[p];return a.createElement.apply(null,o)}return a.createElement.apply(null,n)}u.displayName="MDXCreateElement"},5001:(e,t,n)=>{"use strict";n.r(t),n.d(t,{frontMatter:()=>s,metadata:()=>p,toc:()=>d,default:()=>h});var a,r=n(2122),i=n(9756),o=(n(7294),n(3905)),l=["components"],s={id:"errors",title:"Errors",sidebar_label:"Errors"},p={unversionedId:"errors",id:"errors",isDocsHomePage:!1,title:"Errors",description:"Common Issues",source:"@site/docs/errors.md",sourceDirName:".",slug:"/errors",permalink:"/docs/errors",editUrl:"https://github.com/facebook/pyre-check/tree/master/documentation/website/docs/errors.md",version:"current",sidebar_label:"Errors",frontMatter:{id:"errors",title:"Errors",sidebar_label:"Errors"},sidebar:"pyre",previous:{title:"Types in Python",permalink:"/docs/types-in-python"},next:{title:"Querying Pyre",permalink:"/docs/querying-pyre"}},d=[{value:"Common Issues",id:"common-issues",children:[{value:"Covariance and Contravariance",id:"covariance-and-contravariance",children:[]},{value:"Optional Attributes",id:"optional-attributes",children:[]},{value:"Third-Party Libraries",id:"third-party-libraries",children:[]}]},{value:"Error Codes",id:"error-codes",children:[{value:"0: Unused Ignore",id:"0-unused-ignore",children:[]},{value:"2: Missing Parameter Annotation",id:"2-missing-parameter-annotation",children:[]},{value:"3: Missing Return Annotation",id:"3-missing-return-annotation",children:[]},{value:"4: Missing Attribute Annotation",id:"4-missing-attribute-annotation",children:[]},{value:"5: Missing Global Annotation",id:"5-missing-global-annotation",children:[]},{value:"6: Incompatible Parameter Type",id:"6-incompatible-parameter-type",children:[]},{value:"7: Incompatible Return Type",id:"7-incompatible-return-type",children:[]},{value:"8: Incompatible Attribute Type",id:"8-incompatible-attribute-type",children:[]},{value:"9: Incompatible Variable Type",id:"9-incompatible-variable-type",children:[]},{value:"10: Unbound Name",id:"10-unbound-name",children:[]},{value:"11, 31: Undefined or Invalid Type",id:"11-31-undefined-or-invalid-type",children:[]},{value:"12: Incompatible Awaitable Type",id:"12-incompatible-awaitable-type",children:[]},{value:"13: Uninitialized Attribute",id:"13-uninitialized-attribute",children:[]},{value:"14,15: Behavioral Subtyping",id:"1415-behavioral-subtyping",children:[]},{value:"16: Missing Attributes",id:"16-missing-attributes",children:[]},{value:"17: Incompatible Constructor Annotation",id:"17-incompatible-constructor-annotation",children:[]},{value:"18,21: Undefined Name, Undefined Import",id:"1821-undefined-name-undefined-import",children:[]},{value:"19: Too Many Argument",id:"19-too-many-argument",children:[]},{value:"20: Missing Argument",id:"20-missing-argument",children:[]},{value:"22: Redundant Cast",id:"22-redundant-cast",children:[]},{value:"23: Unable to Unpack",id:"23-unable-to-unpack",children:[]},{value:"24: Invalid Type Parameters",id:"24-invalid-type-parameters",children:[]},{value:"26: Typed Dictionary Access With Non-Literal",id:"26-typed-dictionary-access-with-non-literal",children:[]},{value:"27: Typed Dictionary Key Not Found",id:"27-typed-dictionary-key-not-found",children:[]},{value:"28: Unexpected Keyword",id:"28-unexpected-keyword",children:[]},{value:"29: Call Error",id:"29-call-error",children:[]},{value:"30, 36: Terminating Analysis, Mutually Recursive Type Variables",id:"30-36-terminating-analysis-mutually-recursive-type-variables",children:[]},{value:"31: Invalid Type",id:"31-invalid-type",children:[]},{value:"32: Invalid Argument",id:"32-invalid-argument",children:[]},{value:"33: Prohibited Any",id:"33-prohibited-any",children:[]},{value:"34: Invalid Type Variable",id:"34-invalid-type-variable",children:[]},{value:"35: Illegal Annotation Target",id:"35-illegal-annotation-target",children:[]},{value:"39: Invalid Inheritance",id:"39-invalid-inheritance",children:[]},{value:"40: Invalid Override",id:"40-invalid-override",children:[]},{value:"41: Invalid Assignment",id:"41-invalid-assignment",children:[]},{value:"42: Missing Overload Implementation",id:"42-missing-overload-implementation",children:[]},{value:"43: Incompatible Overload Implementation",id:"43-incompatible-overload-implementation",children:[]},{value:"45: Invalid Class Instantiation",id:"45-invalid-class-instantiation",children:[]},{value:"46: Invalid Type Variance",id:"46-invalid-type-variance",children:[]},{value:"47: Invalid Method Signature",id:"47-invalid-method-signature",children:[]},{value:"48: Invalid Exception",id:"48-invalid-exception",children:[]},{value:"49: Unsafe Cast",id:"49-unsafe-cast",children:[]},{value:"51: Unused Local Mode",id:"51-unused-local-mode",children:[]},{value:"52: Private Protocol Property",id:"52-private-protocol-property",children:[]},{value:"53: Missing Annotation For Captured Variables",id:"53-missing-annotation-for-captured-variables",children:[]},{value:"54: Invalid TypedDict Operation",id:"54-invalid-typeddict-operation",children:[]},{value:"55: TypedDict Initialization Error",id:"55-typeddict-initialization-error",children:[]},{value:"56: Invalid Decoration",id:"56-invalid-decoration",children:[]},{value:"57: Incompatible Async Generator Return Type",id:"57-incompatible-async-generator-return-type",children:[]},{value:"58: Unsupported Operand",id:"58-unsupported-operand",children:[]},{value:"59: Duplicate Type Variables",id:"59-duplicate-type-variables",children:[]},{value:"60: Unable to Concatenate Tuple",id:"60-unable-to-concatenate-tuple",children:[]}]},{value:"Suppression",id:"suppression",children:[{value:"Suppressing Individual Errors",id:"suppressing-individual-errors",children:[]},{value:"Suppressing All Errors",id:"suppressing-all-errors",children:[]},{value:"Suppressing Errors Across Files",id:"suppressing-errors-across-files",children:[]}]}],c=(a="Internal",function(e){return console.warn("Component "+a+" was not imported, exported, or provided by MDXProvider as global scope"),(0,o.kt)("div",e)}),u={toc:d};function h(e){var t=e.components,n=(0,i.Z)(e,l);return(0,o.kt)("wrapper",(0,r.Z)({},u,n,{components:t,mdxType:"MDXLayout"}),(0,o.kt)("h2",{id:"common-issues"},"Common Issues"),(0,o.kt)("h3",{id:"covariance-and-contravariance"},"Covariance and Contravariance"),(0,o.kt)("p",null,(0,o.kt)("a",{parentName:"p",href:"https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)"},"Variance")," is tricky and a common source of confusion for people new to Python's type system."),(0,o.kt)("p",null,"Pyre will error when, for instance, a ",(0,o.kt)("inlineCode",{parentName:"p"},"List[int]")," is passed in when a ",(0,o.kt)("inlineCode",{parentName:"p"},"List[float]")," is expected, as in the following example:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def to_seconds(milliseconds: List[float]) -> List[int]:\n  return [int(x/1000.0) for x in milliseconds]\n\nmy_list: List[int] = [1]\nmy_list = to_seconds(my_list) # Pyre errors here!\n")),(0,o.kt)("p",null,"This code is works perfectly fine at runtime, and we may think that since ",(0,o.kt)("inlineCode",{parentName:"p"},"int")," is a subtype of ",(0,o.kt)("inlineCode",{parentName:"p"},"float")," this should not be a problem for the type checker either. However, consider the following code:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def halve_first_element(list: List[float]) -> None:\n  list[0] /= 2\n\nmy_list: List[int] = [1]\nhalve_first_element(my_list)\nfunction_taking_int(my_list[0]) # Oh no, my_list[0] is 0.5!\n")),(0,o.kt)("p",null,"If we allowed passing in ",(0,o.kt)("inlineCode",{parentName:"p"},"my_list")," to the ",(0,o.kt)("inlineCode",{parentName:"p"},"halve_first_element")," function here, the above code would type check. It's perfectly valid from the perspective of the callee to modify the list's element to be a float, as it was annotated as taking a list of floats, but because this list escapes the scope of the callee, we can't allow this in the type checker."),(0,o.kt)("p",null,"To work around this, we can signal to the type checker that the parameter can't be modified. Here's how you can tell the type checker that you won't change the container in your function:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"# I can't modify milliseconds here, so it's safe to pass a Iterable[int].\ndef to_seconds(milliseconds: Iterable[float]) -> List[int]:\n  return [int(x/1000.0) for x in milliseconds]\n\nmy_list: List[int] = [1]\nmy_list = to_seconds(my_list) # Type checks!\n")),(0,o.kt)("p",null,(0,o.kt)("inlineCode",{parentName:"p"},"typing.Iterable")," is an immutable variant for lists that allows accessing the list without modifying it. Most commonly used generic containers have immutable variants, and I would encourage you to use them for function parameters whenever you don't need to modify a container in your function.\nHere are some immutable variants for commonly used containers:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"typing.List \u2192 typing.Sequence (if you need random access via my_list[id])\ntyping.List \u2192 typing.Iterable (if you're just iterating over the list in a loop and want to support sets as well)\ntyping.Dict \u2192 typing.Mapping\ntyping.Set \u2192 typing.AbstractSet\n")),(0,o.kt)("p",null,"Invariance, combined with type inference, comes with a few gotchas. When you write an expression, Pyre infers the most precise type possible. For instance, Pyre infers the ",(0,o.kt)("inlineCode",{parentName:"p"},"List[int]")," type for ",(0,o.kt)("inlineCode",{parentName:"p"},"[1, 2]"),", even though ",(0,o.kt)("inlineCode",{parentName:"p"},"List[float]")," would be a perfectly valid type here. This can cause issues, as in the following example:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def zeroes(number_of_elements: int) -> List[float]:\n  a = [0] * number_of_elements\n  return a # Pyre errors here!\n")),(0,o.kt)("p",null,"What happened above is that Pyre inferred a type of ",(0,o.kt)("inlineCode",{parentName:"p"},"List[int]")," for a, and invariance kicked in. You can work around this by adding an explicit annotation when declaring a:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def zeroes(number_of_elements: int) -> List[float]:\n  a: List[float] = [0] * number_of_elements\n  return a # Type checks!\n")),(0,o.kt)("h3",{id:"optional-attributes"},"Optional Attributes"),(0,o.kt)("p",null,"A common pattern in Python is to check whether an attribute is ",(0,o.kt)("inlineCode",{parentName:"p"},"None")," before accessing its value. E.g."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"from typing import Optional\n\nclass Data:\n  field: Optional[int]\n\ndef process_field(input: int) -> None:\n  ...\n\ndef process_data(data: Data) -> None:\n  if data.field:\n    # ...\n    process_field(data.field)  # Error: expected `int` but got `Optional[int]`\n")),(0,o.kt)("p",null,"The above fails to type-check because Pyre cannot guarantee that ",(0,o.kt)("inlineCode",{parentName:"p"},"data.field")," is not ",(0,o.kt)("inlineCode",{parentName:"p"},"None")," even after checking explicitly in the line before: ",(0,o.kt)("inlineCode",{parentName:"p"},"field")," could be set to ",(0,o.kt)("inlineCode",{parentName:"p"},"None")," by another thread or it could be a property that returns something different the next time we access it."),(0,o.kt)("p",null,"The preferred way to make this code type-check is to mark the attribute ",(0,o.kt)("inlineCode",{parentName:"p"},"Final"),", i.e. to specify that it can't be reassigned."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"from typing import Final, Optional\n\nclass Data:\n  # Needs to be assigned in the constructor and cannot be changed afterwards.\n  field: Final[Optional[int]] = 1\n")),(0,o.kt)("p",null,"It is always safe to refine attributes when their types are ",(0,o.kt)("inlineCode",{parentName:"p"},"Final"),"."),(0,o.kt)("p",null,"Alternatively, it is also safe to assign the attribute to a local variable before accessing its value."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def process_data(data: Data) -> None:\n  field = data.field\n  if field:\n    process_field(field)\n")),(0,o.kt)("h3",{id:"third-party-libraries"},"Third-Party Libraries"),(0,o.kt)("p",null,"Not all third-party libraries come with Python code that Pyre can analyze (e.g. ",(0,o.kt)("inlineCode",{parentName:"p"},"cython")," modules), and some libraries contain source code without annotations. This will often show up in the form of undefined attribute errors:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"Undefined attribute [16]: Module <library> has no attribute <some attribute>.\n")),(0,o.kt)("p",null,"Since it is not always possible to annotate code, PEP 484 specifies a format for ",(0,o.kt)("a",{parentName:"p",href:"https://www.python.org/dev/peps/pep-0484/#stub-files"},"stub files")," with a ",(0,o.kt)("inlineCode",{parentName:"p"},".pyi")," extension. Pyre will look for stub files in ",(0,o.kt)("a",{parentName:"p",href:"https://github.com/python/typeshed"},"typeshed"),", or next to your source code. You can also provide additional paths to Pyre to look for stubs (see ",(0,o.kt)("a",{parentName:"p",href:"/docs/configuration"},"Configuration"),")."),(0,o.kt)(c,{mdxType:"Internal"}),(0,o.kt)("h2",{id:"error-codes"},"Error Codes"),(0,o.kt)("p",null,"Different errors raised by Pyre have different error codes. E.g. in"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-bash"},"(venv) $ pyre\n \u019b Found 1 type error!\ntest.py:1:0 Incompatible variable type [9]: a is declared to have type `int` but is used as type `str`.\n")),(0,o.kt)("p",null,"The ",(0,o.kt)("inlineCode",{parentName:"p"},"9")," in the brackets indicates that we raised an error with code 9."),(0,o.kt)("h3",{id:"0-unused-ignore"},"0: Unused Ignore"),(0,o.kt)("p",null,"Pyre fixmes and ignores allow you to ignore specific type errors by their code until you are able to fix them. In order to avoid outdated fixme comments in your project, Pyre will also error when a fixme is no longer needed. Removing the fixme comment will resolve the error."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"# pyre-fixme[7] # unused ignore\ndef foo() -> int:\n  return 1\n")),(0,o.kt)("h3",{id:"2-missing-parameter-annotation"},"2: Missing Parameter Annotation"),(0,o.kt)("p",null,"If strict mode is turned on, Pyre will error when a function argument is either annotated with a return type that contains ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.Any"),", or is not annotated with any type at all (in which case Pyre will treat it as ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.Any")," by default). It will also error when a method argument is not annotated, unless that argument is the first argument of a bound or static method (i.e. ",(0,o.kt)("inlineCode",{parentName:"p"},"self"),", whose type pyre can infer)."),(0,o.kt)("p",null,"We enforce typed argument because ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.Any")," can hide type errors that will happen at runtime:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import Any\n\ndef say_hello(name) -> None:\n    print("Hello " + name)\n\n\n# This line will raise at runtime, but no type error since `say_hello`s `name` has type `Any`.\nsay_hello(42)\n')),(0,o.kt)("p",null,"You can silence this by adding a non-",(0,o.kt)("inlineCode",{parentName:"p"},"Any")," annotation to all parameters of functions and methods (other than ",(0,o.kt)("inlineCode",{parentName:"p"},"self")," and ",(0,o.kt)("inlineCode",{parentName:"p"},"cls")," for bound and class methods, which you may omit)."),(0,o.kt)("h3",{id:"3-missing-return-annotation"},"3: Missing Return Annotation"),(0,o.kt)("p",null,"If strict mode is turned on, Pyre will error when a function is either annotated with a return type that contains ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.Any"),", or is not annotated with any return type at all (in which case Pyre will treat it as returning ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.Any")," by default)."),(0,o.kt)("p",null,"This is bad because a return type of ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.Any")," may potentially hiding legitimate type errors that may happen at runtime:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import Any\n\ndef f():\n  return 42\n\n# This line will raise at runtime, but no type error here since `f()` has type `Any`.\nprint("a" + f())\n')),(0,o.kt)("p",null,"The best way to silence this error is to add non-",(0,o.kt)("inlineCode",{parentName:"p"},"Any")," return annotation to every function."),(0,o.kt)("h3",{id:"4-missing-attribute-annotation"},"4: Missing Attribute Annotation"),(0,o.kt)("p",null,"In strict mode, Pyre will error when an attribute does not have an annotation."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class A:\n    b = foo() # Missing attribute annotation\n")),(0,o.kt)("p",null,"Adding a type annotation will resolve this error."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class A:\n  b: int = foo()\n")),(0,o.kt)("h3",{id:"5-missing-global-annotation"},"5: Missing Global Annotation"),(0,o.kt)("p",null,"If strict mode is turned on, Pyre will error when a globally accessible variable is not annotated. If pyre was able to infer a type for the\nvariable, it will emit this type in the error message. The fix is usually to add an annotation to the variable."),(0,o.kt)("p",null,"Note: This error has also arisen when there is some ambiguity of whether a declaration is a global expression or a type alias, in these cases pyre assumes it is an expression. Adding a ",(0,o.kt)("inlineCode",{parentName:"p"},": TypeAlias")," annotation lets pyre know that it is a type alias and solves the problem."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing_extensions import TypeAlias\n\n# This declaration would result in an error\nMyTypeAlias = Dict[str, "AnotherTypeAlias"]\n\n# This declaration ensures that pyre knows MyTypeAlias is a type alias\nMyTypeAlias: TypeAlias = Dict[str, "AnotherTypeAlias"]\n')),(0,o.kt)("h3",{id:"6-incompatible-parameter-type"},"6: Incompatible Parameter Type"),(0,o.kt)("p",null,"Pyre will error if an argument passed into a function call does not match the expected parameter type of that function."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def takes_int(x: int) -> None:\n  pass\ndef f(x: Optional[int]) -> None:\n  takes_int(x) # Incompatible parameter type error\n")),(0,o.kt)("p",null,"If you are seeing errors with invariant containers where some ",(0,o.kt)("inlineCode",{parentName:"p"},"Container[T]")," is expected but you are passing ",(0,o.kt)("inlineCode",{parentName:"p"},"Container[S]")," where ",(0,o.kt)("inlineCode",{parentName:"p"},"S < T"),", please see ",(0,o.kt)("a",{parentName:"p",href:"/docs/errors#covariance-and-contravariance"},"Covariance and Contravariance"),"."),(0,o.kt)("h3",{id:"7-incompatible-return-type"},"7: Incompatible Return Type"),(0,o.kt)("p",null,"Pyre will error when the value returned from a function does not match the annotation."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'def foo() -> int:\n  return "" # incompatible return type\n')),(0,o.kt)("p",null,"Updating the return annotation, or the value returned from the function will resolve this error."),(0,o.kt)("h3",{id:"8-incompatible-attribute-type"},"8: Incompatible Attribute Type"),(0,o.kt)("p",null,"Pyre will error if a value is assigned to an attribute that does not match the annotated type of that attribute."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'class Foo:\n  x: int = 0\n\ndef f(foo: Foo) -> None:\n  foo.x = "abc" # Incompatible attribute type error\n')),(0,o.kt)("p",null,"If you are seeing errors with invariant containers where some ",(0,o.kt)("inlineCode",{parentName:"p"},"Container[T]")," is expected but you are passing ",(0,o.kt)("inlineCode",{parentName:"p"},"Container[S]")," where ",(0,o.kt)("inlineCode",{parentName:"p"},"S < T"),", please see ",(0,o.kt)("a",{parentName:"p",href:"/docs/errors#covariance-and-contravariance"},"Covariance and Contravariance"),"."),(0,o.kt)("h3",{id:"9-incompatible-variable-type"},"9: Incompatible Variable Type"),(0,o.kt)("p",null,"Pyre will error when assigning incompatible types to local variables and parameters that were explicitly annotated."),(0,o.kt)("p",null,"That is, the following will error:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'def f(x: int) -> None:\n  x = "" # Incompatible variable type error\n  y: int = 1\n  y = "" # Incompatible variable type error\n')),(0,o.kt)("p",null,"The rationale here is that it's surprising for an explicitly annotated variable to have an\nincompatible type later on in the same function."),(0,o.kt)("p",null,"If you are constructing an object that is generic over an invariant type, you may run into an error:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"_T = TypeVar('_T')\n\nclass Foo(Generic[_T]):\n    def __init__(self, x: _T) -> None: ...\n\ndef f() -> None:\n    foo: Foo[Optional[int]] = Foo(x=1) # Incompatible variable type error\n")),(0,o.kt)("p",null,"This is due to the fact that ",(0,o.kt)("inlineCode",{parentName:"p"},"Foo[X]")," is not less than ",(0,o.kt)("inlineCode",{parentName:"p"},"Foo[Y]")," even if ",(0,o.kt)("inlineCode",{parentName:"p"},"X < Y")," when the type variable is invariant.\nYou can declare your intention to initialize the object with a wider type than is given to fix this error:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def f() -> None:\n    foo: Foo[Optional[int]] = Foo[Optional[int]](x=1)\n")),(0,o.kt)("h3",{id:"10-unbound-name"},"10: Unbound Name"),(0,o.kt)("p",null,"Pyre produces an Unbound Name error when your code access a variable (local or global) that pyre believes is not defined."),(0,o.kt)("p",null,"In most cases code that does this is invalid and will always fail. For example this code will always fail at runtime:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def f() -> int:\n    return x  # use of an unbound name x\n")),(0,o.kt)("p",null,"There are some cases where python code that works fine at runtime could produce this error, for example if a function implicitly sets a module-level global variable that is not declared in the toplevel. Pyre will not accept this because module-level globals require type annotations, and if they have no declaration there is nowhere to put the annotation:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def set_x() -> None:\n    global x\n    x = 42\n\ndef use_x() -> None:\n    print(x)\n\n\n# this code will run fine, but pyre cannot analyze the type or use of the\n# implicitly-defined global x and will complain about an unbound name.\nset_x()\nuse_x()\n")),(0,o.kt)("p",null,"You can fix this by explicitly adding a declaration of the top-level variable ",(0,o.kt)("inlineCode",{parentName:"p"},"x"),", for example:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"x : Optional[int] = None\n")),(0,o.kt)("h3",{id:"11-31-undefined-or-invalid-type"},"11, 31: Undefined or Invalid Type"),(0,o.kt)("p",null,"Pyre recognizes class names as valid annotations. Most basic types are imported from the ",(0,o.kt)("inlineCode",{parentName:"p"},"typing")," module or are already available from builtins like ",(0,o.kt)("inlineCode",{parentName:"p"},"str"),", ",(0,o.kt)("inlineCode",{parentName:"p"},"int"),", ",(0,o.kt)("inlineCode",{parentName:"p"},"bool"),", etc. You can also define your own type alias on the global scope, which can be used as annotations:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"from typing_extensions import TypeAlias\n\nINT_OR_STR: TypeAlias = Union[int, str]\n")),(0,o.kt)("p",null,"If you use a name as an annotation that is not a valid type or valid alias, you will see this error:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'GLOBAL_VALUE = "string"\n\ndef f0() -> GLOBAL_VALUE: ... # Error! `GLOBAL_VALUE` is a value, not a type.\n\ndef f1() -> type(GLOBAL_VALUE): ...   # Error! Static type annotations cannot be dynamically computed.\n\ndef f2() -> [int]: ...  # Error! `[int]` is not a valid type. If you mean a list of int, use `typing.List[int]`.\n\ndef f3() -> (int, str): ...  # Error! `(int, str)` is not a valid type. If you mean a pair of int and str, use `typing.Tuple[int, str]`.\n\nfrom typing import Callable\ndef f4() -> Callable[[int]]: ...  # Error! `Callable[[int]]` is not a valid type because the return type of the callable is missing. Good example: `Callable[[int], int]`.\n\ndef f5() -> Callable[int, int]: ...  # Error! `Callable[int, int]` is not a valid type. The parameter types of the callable must be enclosed in square brackets. Good example: `Callable[[int], int]`.\n\nfrom typing_extensions import Final\ndef f6() -> List[Final[int]]: ...  # Error! `Final` may only be used as the outermost type in annotations. See PEP 591.\n\nfrom typing_extensions import Literal\ndef f7() -> Literal[GLOBAL_VALUE]: ...  # Error! Only literals are allowed as parameters for `Literal`. See PEP586. Good example: `Literal[42]` or `Literal["string"]`.\n')),(0,o.kt)("p",null,"You can fix this error by verifying that your annotation is"),(0,o.kt)("ol",null,(0,o.kt)("li",{parentName:"ol"},"statically determined."),(0,o.kt)("li",{parentName:"ol"},"properly imported from ",(0,o.kt)("inlineCode",{parentName:"li"},"typing")," if applicable."),(0,o.kt)("li",{parentName:"ol"},"properly defined in the module you are importing from. If the module you are importing from has a ",(0,o.kt)("a",{parentName:"li",href:"/docs/errors#third-party-libraries"},"stub file"),", you should check the definition there."),(0,o.kt)("li",{parentName:"ol"},"properly adhere to the additional rules of special types (e.g. ",(0,o.kt)("inlineCode",{parentName:"li"},"Callable"),", ",(0,o.kt)("inlineCode",{parentName:"li"},"Final"),", and ",(0,o.kt)("inlineCode",{parentName:"li"},"Literal"),").")),(0,o.kt)("p",null,"For type aliases, check that your type alias is defined"),(0,o.kt)("ol",null,(0,o.kt)("li",{parentName:"ol"},"with a valid type on the RHS. If you provide an annotation for the TypeAlias assignment, it must be ",(0,o.kt)("inlineCode",{parentName:"li"},"typing_extensions.TypeAlias"),"."),(0,o.kt)("li",{parentName:"ol"},"on the global scope, not nested inside a function or class.")),(0,o.kt)("h3",{id:"12-incompatible-awaitable-type"},"12: Incompatible Awaitable Type"),(0,o.kt)("p",null,"In strict mode, pyre will verify that all calls to ",(0,o.kt)("inlineCode",{parentName:"p"},"await")," are on awaitable values, to ensure that you cannot get a runtime error awaiting an object that is not a coroutine."),(0,o.kt)("p",null,"A common situation where working code will produce this error is when pyre cannot statically verify that an awaitable is non-Null, for example:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"import asyncio\n\nasync def f(flag: bool) -> None:\n    if flag:\n        task = asyncio.create_task(asyncio.sleep(1))\n    else:\n        task = None\n    await task  # would throw a ValueError if flag were false\n\nasyncio.run(f(True))\n")),(0,o.kt)("p",null,"In this example, ",(0,o.kt)("inlineCode",{parentName:"p"},"task")," will always be a valid awaitable unless some other module overwrites the global ",(0,o.kt)("inlineCode",{parentName:"p"},"flag"),", but pyre cannot prove that this does not happen. The error we get has the message"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"Expected an awaitable but got `typing.Optional[asyncio.tasks.Task[None]]`\n")),(0,o.kt)("p",null,"You can fix this error by ensuring that the awaited object has an awaitable type. In the case of optional values, you can use refinement to rule out ",(0,o.kt)("inlineCode",{parentName:"p"},"None"),". The example above can be fixed by tweaking the definition of ",(0,o.kt)("inlineCode",{parentName:"p"},"main"),":"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"async def f(flag: bool) -> None:\n    if flag:\n        task = asyncio.create_task(asyncio.sleep(1))\n    else:\n        task = None\n    if task is not None:\n      await task\n")),(0,o.kt)("h3",{id:"13-uninitialized-attribute"},"13: Uninitialized Attribute"),(0,o.kt)("p",null,"In strict mode, pyre will throw an error for class attributes which are declared without default values if they are not initialized in a constructor, for example:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class A:\n    x : int\n\n    def __init__(self) -> None:\n        pass\n")),(0,o.kt)("p",null,"For a case like this, you can fix the error either by setting a default value like ",(0,o.kt)("inlineCode",{parentName:"p"},"x : int = 0")," at the class level, or by setting ",(0,o.kt)("inlineCode",{parentName:"p"},"x")," in the constructor e.g. ",(0,o.kt)("inlineCode",{parentName:"p"},"self.x = 0"),"."),(0,o.kt)("h4",{id:"dataclass-like-classes"},"Dataclass-like classes"),(0,o.kt)("p",null,'One case where this can occur is when using a library providing a "dataclass-like" decorator that, for example, autogenerates a constructor setting attributes.'),(0,o.kt)("p",null,"Pyre currently knows that that uninitialized attributes of classes wrapped in ",(0,o.kt)("inlineCode",{parentName:"p"},"dataclass")," and ",(0,o.kt)("inlineCode",{parentName:"p"},"attrs")," decorators will generate constructors that set the attributes. But it does not understand many custom libraries that do similar things, for example test frameworks, or new decorators that wrap the dataclass decorator and add more logic."),(0,o.kt)("p",null,"There is not currently a way to fix this other than via ",(0,o.kt)("inlineCode",{parentName:"p"},"pyre-ignore")," or ",(0,o.kt)("inlineCode",{parentName:"p"},"pyre-fixme")," directives. The python typing community is aware of this problem but has not yet settled on a solution, you can see discussion ",(0,o.kt)("a",{parentName:"p",href:"https://mail.python.org/archives/list/typing-sig@python.org/thread/TXL5LEHYX5ZJAZPZ7YHZU7MVFXMVUVWL/"},"here"),"."),(0,o.kt)("h3",{id:"1415-behavioral-subtyping"},"14,15: Behavioral Subtyping"),(0,o.kt)("p",null,"Method overrides should follow\n",(0,o.kt)("a",{parentName:"p",href:"https://en.wikipedia.org/wiki/Liskov_substitution_principle"},"Liskov's substitution principle"),".\nIn short, parameter types can't be more restrictive and return types\ncan't be more permissive in overridden methods. To see why, consider the following example:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def width(image: Image) -> float:\n    return image.width()\n")),(0,o.kt)("p",null,"Say we now have different implementations of our ",(0,o.kt)("inlineCode",{parentName:"p"},"Image")," class, one of which\nviolates the substitution principle:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class Image:\n    def width() -> float: pass\n\nclass JpegImage(Image):\n    def width() -> int: return 10  # this is fine\n\nclass ComplexImage(Image):\n    def width() -> complex: return 1j\n\ndef foo() -> None:\n    image: Image = ComplexImage()\n    print(int(image.width()))\n")),(0,o.kt)("p",null,"The above code fails at runtime with ",(0,o.kt)("inlineCode",{parentName:"p"},"TypeError: can't convert complex to int"),". The case for parameters follows analogously."),(0,o.kt)("h4",{id:"common-reasons"},"Common Reasons"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},(0,o.kt)("inlineCode",{parentName:"p"},"Could not find parameter y in overriding signature."),": Check if the overriding function can accept all arguments that the overridden function can."),(0,o.kt)("pre",{parentName:"li"},(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class Base:\n    def foo(self, x: int, y: str) -> None:\n        pass\n\nclass Child(Base):\n    def foo(self, x: int) -> None:\n        pass\n"))),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},(0,o.kt)("inlineCode",{parentName:"p"},"Type Foo is not a subtype of the overridden attribute type Bar"),":"),(0,o.kt)("pre",{parentName:"li"},(0,o.kt)("code",{parentName:"pre",className:"language-python"},'class Base:\n    a: int = 0\nclass Child(Base):\n    a: str = ""\n\ndef foo() -> None:\n    base: Base = Child()\n    base.a + 1\n')),(0,o.kt)("p",{parentName:"li"},"This would fail at runtime with ",(0,o.kt)("inlineCode",{parentName:"p"},'TypeError: can only concatenate str (not "int") to str'),".")),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},(0,o.kt)("inlineCode",{parentName:"p"},"Returned type Foo is not a subtype of the overridden return Bar."),": Check for reasons like ",(0,o.kt)("a",{parentName:"p",href:"/docs/errors#covariance-and-contravariance"},"invariance"),"."))),(0,o.kt)("h3",{id:"16-missing-attributes"},"16: Missing Attributes"),(0,o.kt)("p",null,"Your code is most likely trying to access an attribute that Pyre does not know about.\nPyre has various ways of inferring what is an attribute of an object:"),(0,o.kt)("h4",{id:"explicitly-declare-the-attribute"},"Explicitly Declare the Attribute"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class Derp:\n    my_attribute: int = 1\n\n    @property\n    def my_property(self) -> int: ...\n")),(0,o.kt)("h4",{id:"implicitly-declare-the-attribute"},"Implicitly Declare the Attribute"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class Derp:\n    def __init__(self, foo: str) -> None:\n        self.my_attribute: int = 1\n\n        # The `foo` attribute is inferred to have type `str` because the\n        # parameter `foo` has type `str`.\n        self.foo = foo\n")),(0,o.kt)("p",null,"Pyre does one level of inlining to infer implicit parameters\nWe suggest you do not heavily rely on this feature as it is not sound and makes our code brittle.\nSupport for this is temporary."),(0,o.kt)("h4",{id:"common-reasons-1"},"Common Reasons"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},(0,o.kt)("inlineCode",{parentName:"p"},"Optional type has no attribute foo."),": See ",(0,o.kt)("a",{parentName:"p",href:"/docs/errors#optional-attributes"},"Optional attributes"),".")),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},(0,o.kt)("inlineCode",{parentName:"p"},"Foo has no attribute bar."),": Check if you have explicitly provided the type for ",(0,o.kt)("inlineCode",{parentName:"p"},"bar")," either in the constructor or as a class attribute.")),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},(0,o.kt)("inlineCode",{parentName:"p"},"Module foo has no attribute bar"),": Check if the library has ",(0,o.kt)("a",{parentName:"p",href:"/docs/errors#third-party-libraries"},"stubs"),". If so, you may need to add the function, class, or global variable to the stub.")),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"A library class has an attribute but it is not recognized by Pyre: Check if the library has ",(0,o.kt)("a",{parentName:"p",href:"/docs/errors#third-party-libraries"},"stubs"),". If so, you may need to add the attribute to the class in the stub.")),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"Your class has dynamic attributes: Consider using ",(0,o.kt)("inlineCode",{parentName:"p"},"__getattr__")," in a ",(0,o.kt)("a",{parentName:"p",href:"/docs/types-in-python#when-source-code-is-not-available"},"stub")," so that Pyre doesn't complain about those attributes."))),(0,o.kt)("h3",{id:"17-incompatible-constructor-annotation"},"17: Incompatible Constructor Annotation"),(0,o.kt)("p",null,(0,o.kt)("a",{parentName:"p",href:"https://www.python.org/dev/peps/pep-0484/#the-meaning-of-annotations"},"PEP 484")," specifies that ",(0,o.kt)("inlineCode",{parentName:"p"},"__init__")," method of any class must be annotated to return ",(0,o.kt)("inlineCode",{parentName:"p"},"None"),". Pyre will emit an error if the user's annotation does not conform to the specification."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'# pyre-strict\n\nclass A:\n    def __init__(self) -> "A":  # Error 17: Invalid return annotation of `__init__`.\n        ...\n\nclass B:\n    def __init__(self) -> None:  # OK\n        ...\n')),(0,o.kt)("h3",{id:"1821-undefined-name-undefined-import"},"18,21: Undefined Name, Undefined Import"),(0,o.kt)("p",null,'Error 18 ("Undefined name") is raised when your code tries to access a variable or function that Pyre could not resolve.\nThis is usually caused by failing to import the proper module.'),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"  # 'import some_module' is missing\n  some_module.some_func()\n")),(0,o.kt)("p",null,'Pyre will raise error 21 instead ("Undefined import") when the import statement is present, but the module to be imported could not be found in the search path.\nIf the module provides stub files, please provide their location via the ',(0,o.kt)("inlineCode",{parentName:"p"},"--search-path")," commandline parameter."),(0,o.kt)("h3",{id:"19-too-many-argument"},"19: Too Many Argument"),(0,o.kt)("p",null,"Pyre verifies that you pass a legal number of arguments to functions."),(0,o.kt)("p",null,"The most obvious way to encounter this error is to just pass too many arguments to a function:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def f(x: int) -> int:\n    return x\n\nf(5, 6)  # this would throw a TypeError at runtime, and pyre complains\n")),(0,o.kt)("p",null,"To fix this, make sure you pass the correct number of parameters. In some cases you may encounter this error if you intended to use a variadic argument (",(0,o.kt)("inlineCode",{parentName:"p"},"*args"),") or to set a default value."),(0,o.kt)("p",null,"Pyre will also throw this error if you pass too many positional arguments to\na function that uses python's ability restrict arguments to be keyword-only\nspecified by ",(0,o.kt)("a",{parentName:"p",href:"https://www.python.org/dev/peps/pep-3102/"},"PEP 3102"),":"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def f(*, x: int) -> int:\n    return x\n\nf(5)  # As before, this throws a TypeError because x is positional-only\nf(x=5)  # this line will typecheck and run without error\n")),(0,o.kt)("h3",{id:"20-missing-argument"},"20: Missing Argument"),(0,o.kt)("p",null,"Pyre verifies that function calls provide all the expected arguments, so it will complain about code like this:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def f(x: int) -> ing:\n    return x\n\nf()\n")),(0,o.kt)("p",null,"To fix this, make sure all required arguments are provided."),(0,o.kt)("h3",{id:"22-redundant-cast"},"22: Redundant Cast"),(0,o.kt)("p",null,"Pyre will warn when you attempt to use ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.cast")," to cast a variable to a type that the type checker already knows that variable has. This is because ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.cast")," is purely a tool for communicating with the static type checker, and will not provide any runtime guarantees. Therefore a redundant cast provides no value and is likely a mistake."),(0,o.kt)("p",null,"If you are trying to document the type of the variable, you can provide an explicit annotation where it is declared. If you are trying to add a sanity check at runtime that the type of a variable is what you already believe it must be, use ",(0,o.kt)("inlineCode",{parentName:"p"},"isinstance"),"."),(0,o.kt)("h3",{id:"23-unable-to-unpack"},"23: Unable to Unpack"),(0,o.kt)("p",null,"Pyre will warn you when trying to assign a value to a tuple with the wrong number of items."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def foo() -> None:\n    a, b = (1, 2, 3)\n    x, y = 42\n\n$ pyre\nUnable to unpack [23]: Unable to unpack 3 values, 2 were expected.\nUnable to unpack [23]: Unable to unpack `int` into 2 values.\n")),(0,o.kt)("p",null,"Common reasons:"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"Trying to assign an ",(0,o.kt)("inlineCode",{parentName:"p"},"Optional")," value to a tuple:"),(0,o.kt)("pre",{parentName:"li"},(0,o.kt)("code",{parentName:"pre",className:"language-python"},'def bar() -> None:\n    x = None\n    if 2 + 2 == 4:\n        x = ("a", "b")\n\n    a, b = x\n\n$ pyre\nUnable to unpack [23]: Unable to unpack `typing.Optional[typing.Tuple[str, str]]` into 2 values.\n'))),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"Unpacking an incorrect number of elements when looping over a list:"),(0,o.kt)("pre",{parentName:"li"},(0,o.kt)("code",{parentName:"pre",className:"language-python"},"for a, b in [1, 2, 3]:\n    print(a, b)\n\n$ pyre\nUnable to unpack [23]: Unable to unpack `int` into 2 values.\n")))),(0,o.kt)("h3",{id:"24-invalid-type-parameters"},"24: Invalid Type Parameters"),(0,o.kt)("p",null,"Pyre will error if a generic type annotation is given with unexpected type parameters."),(0,o.kt)("h4",{id:"generic-type-expects-x-type-parameters-"},'"Generic type expects X type parameters ..."'),(0,o.kt)("p",null,"Either too few or too many type parameters were provided for the container type. For example,"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"x: List[int, str] = [] # Invalid type parameters error\n")),(0,o.kt)("p",null,"In this case, ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.List")," is a generic type taking exactly one type parameter. If we pass a single parameter, this resolves the issue."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"x: List[Union[int, str]] = [] # No error\n")),(0,o.kt)("p",null,"If you do not know or do not want to specify the type parameters, use ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.Any")," but still ensure the arity is correct."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"x: List = [] # Invalid type parameters error\nx: List[Any] = [] # No error\n")),(0,o.kt)("p",null,"Note: You may see a suggestion to use ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.List")," instead of builtins ",(0,o.kt)("inlineCode",{parentName:"p"},"list")," as the type annotation when providing type parameters. This is to avoid runtime errors, because the builtin ",(0,o.kt)("inlineCode",{parentName:"p"},"list")," does not support subscripting and ",(0,o.kt)("inlineCode",{parentName:"p"},"list[int]")," is therefore not runtime-friendly."),(0,o.kt)("h4",{id:"non-generic-type-cannot-take-type-parameters-"},'"Non-generic type cannot take type parameters ..."'),(0,o.kt)("p",null,"Type parameters are only meaningful if the container type is generic. Passing in the type parameter binds the provided parameter type to the generic in the container class. For example,"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class Container:\n  def add(element: int) -> None: ...\n  def get_element() -> int: ...\n\nx: Container[int] = Container() # Invalid type parameter error\n")),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"T = TypeVar('T')\n\nclass Container(Generic[T]):\n  def add(element: T) -> None: ...\n  def get_element() -> T: ...\n\nx: Container[int] = Container()\nx.get_element() # returns int\n\ny: Container[str] = Container()\ny.get_element() # returns str\n")),(0,o.kt)("h4",{id:"type-parameter-violates-constraints-"},'"Type parameter violates constraints ..."'),(0,o.kt)("p",null,"If a container class is generic over a type variable with given type bounds, any type parameter used must comply with those type bounds. For example,"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"T = TypeVar('T', bound=Union[int, bool])\n\nclass Container(Generic[T]):\n  def add(element: T) -> None: ...\n  def get_element() -> T: ...\n\nx: Container[int] = Container() # No error\n\ny: Container[str] = Container() # Invalid type parameter error\n")),(0,o.kt)("h3",{id:"26-typed-dictionary-access-with-non-literal"},"26: Typed Dictionary Access With Non-Literal"),(0,o.kt)("p",null,"In python, ",(0,o.kt)("a",{parentName:"p",href:"https://www.python.org/dev/peps/pep-0589/#id21"},"typed dictionaries can only be accessed using literal strings")," that can be statically verified as valid. As a result, code like this will not typecheck even though it works at runtime, because we cannot statically verify that ",(0,o.kt)("inlineCode",{parentName:"p"},"key")," in ",(0,o.kt)("inlineCode",{parentName:"p"},"print_value")," is a valid ",(0,o.kt)("inlineCode",{parentName:"p"},"Shape")," key:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import TypedDict\n\nclass Shape(TypedDict):\n    sides: int\n    color: str\n\nshape: Shape = {"sides": 4, "color": "blue"}\n\nprint(shape["sides"])  # this is fine because "sides" is a literal\n\nfor key in ["sides", "color"]:\n    print(key, shapes[key]) . # pyre will complain here because it can\'t prove `key` is valid\n')),(0,o.kt)("p",null,"The example above shows a situation where you might hit this error: when you want to iterate over the fields of a typed dict. A suggested fix is to use type-safe operations like ",(0,o.kt)("inlineCode",{parentName:"p"},"dictionary.items")," instead. For example the following code produces the same results but typechecks:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"for key, value  in shapes.items():\n    print(key, value)\n")),(0,o.kt)("p",null,"In other cases where you need to access a ",(0,o.kt)("inlineCode",{parentName:"p"},"TypedDict")," using a variable as a key, you can use ",(0,o.kt)("inlineCode",{parentName:"p"},"dictionary.get(key)"),"."),(0,o.kt)("h3",{id:"27-typed-dictionary-key-not-found"},"27: Typed Dictionary Key Not Found"),(0,o.kt)("p",null,"If you try to access a typed dictionary with a string literal that pyre knows is not a valid key, pyre will emit an error:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import TypedDict\n\nclass Shape(TypedDict):\n    sides: int\n    color: str\n\ndef f(shape: Shape) -> None:\n    print(shape["location"])  # error here\n')),(0,o.kt)("p",null,"A possible fix: pyre considers instances of any ",(0,o.kt)("inlineCode",{parentName:"p"},"TypedDict")," with additional fields to be a subtype of ",(0,o.kt)("inlineCode",{parentName:"p"},"Shape"),", so in many cases you could handle the need for a ",(0,o.kt)("inlineCode",{parentName:"p"},"location")," field in some ",(0,o.kt)("inlineCode",{parentName:"p"},"Shape")," dicts by creating a new type as follows:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import TypedDict\n\nclass Shape(TypedDict):\n    sides: int\n    color: str\n\nclass ShapeWithLocation(TypedDict):\n    sides: int\n    color: str\n    location: str\n\ndef f(shape: ShapeWithLocation) -> None:\n    print(shape["location"])  # okay\n    g(shape)  # also okay: ShapeWithLocation is a subtype of Shape\n\ndef g(shape: Shape) -> None:\n    print(shape)\n')),(0,o.kt)("h3",{id:"28-unexpected-keyword"},"28: Unexpected Keyword"),(0,o.kt)("p",null,"Pyre will error if attempting to pass an argument by name and there are no parameters with a matching name. For example,"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'def foo(integer: int, string: str) -> None: ...\n\nfoo(1, "one")  # no error\nfoo(string="one", integer=1)  # no error\nfoo(integer=1, undefined="one")  # type error\n')),(0,o.kt)("h3",{id:"29-call-error"},"29: Call Error"),(0,o.kt)("p",null,"Pyre will emit an error on seeing a call of one of the following types:"),(0,o.kt)("ol",null,(0,o.kt)("li",{parentName:"ol"},(0,o.kt)("p",{parentName:"li"},"The called object is not a function. This means that its inferred type is not Callable and it is not an instance of a class which implements a ",(0,o.kt)("inlineCode",{parentName:"p"},"__call__")," method. This could happen due to user error (the object is indeed not a function) or due to an incorrect or incomplete type stub for the object's class causing pyre to infer the wrong type.")),(0,o.kt)("li",{parentName:"ol"},(0,o.kt)("p",{parentName:"li"},"The call cannot be safely typed since the types and kinds of its parameters depend on a type variable. This is seen when the callable is typed using a ParameterSpecification type variable and the ",(0,o.kt)("inlineCode",{parentName:"p"},"*args")," and ",(0,o.kt)("inlineCode",{parentName:"p"},"**kwargs")," are not passed into the call correctly, i.e. together and in order. (For more details see ",(0,o.kt)("a",{parentName:"p",href:"https://www.python.org/dev/peps/pep-0612/#the-components-of-a-paramspec"},"PEP 612"),")"))),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from pyre_extensions import ParameterSpecification\n\nP = ParameterSpecification("P")\n\ndef decorator(f: Callable[P, int]) -> Callable[P, None]:\n\n  def foo(*args: P.args, **kwargs: P.kwargs) -> None:\n\n    f(*args, **kwargs)    # Accepted, should resolve to int\n\n    f(*args)              # Rejected\n\n    f(*kwargs, **args)    # Rejected\n\n    f(1, *args, **kwargs) # Rejected\n\n  return foo\n')),(0,o.kt)("h3",{id:"30-36-terminating-analysis-mutually-recursive-type-variables"},"30, 36: Terminating Analysis, Mutually Recursive Type Variables"),(0,o.kt)("p",null,"Either of these errors indicates a bug in Pyre. Please open an issue on ",(0,o.kt)("a",{parentName:"p",href:"https://github.com/facebook/pyre/issues"},"Github"),"."),(0,o.kt)("h3",{id:"31-invalid-type"},"31: Invalid Type"),(0,o.kt)("p",null,"This indicates that you are using some expression that pyre does not understand as a type."),(0,o.kt)("p",null,"Some situations where you might run into this:"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"Using a list of types rather than ",(0,o.kt)("inlineCode",{parentName:"p"},"List[type]"),":"),(0,o.kt)("pre",{parentName:"li"},(0,o.kt)("code",{parentName:"pre",className:"language-python"},'x: [str] = ["a string"]\n')),(0,o.kt)("p",{parentName:"li"},"You can fix this by using the ",(0,o.kt)("inlineCode",{parentName:"p"},"List")," type:"),(0,o.kt)("pre",{parentName:"li"},(0,o.kt)("code",{parentName:"pre",className:"language-python"},'x: list[str] = ["a string"]\n'))),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"Using a constructor call rather than a bare class name:"),(0,o.kt)("pre",{parentName:"li"},(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class A:\n    ...\n\na: A() = A()\n")),(0,o.kt)("p",{parentName:"li"},"You can fix this by using a bare type name:"),(0,o.kt)("pre",{parentName:"li"},(0,o.kt)("code",{parentName:"pre",className:"language-python"},"a: A = A()\n")))),(0,o.kt)("h3",{id:"32-invalid-argument"},"32: Invalid Argument"),(0,o.kt)("p",null,"This error usually means you are using a variable in a way that is incompatible with its structure, either as an argument to a function call or as part of a data structure."),(0,o.kt)("p",null,'This could be from using an invalid variadic parameter (informally a "splat"):'),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"x: int = 5\n\nprint(*x)   # invalid use of x, which is not iterable\n")),(0,o.kt)("p",null,'or using an invalid keyword parameter (informally a "double-splat"):'),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"import typing\n\nx: int = 5\n\nd: typing.Dict[int, int] = {**x}  # invalid use of x, which is not a mapping\n\ndict(**d)  # invalid use of d; function kwargs must be a mapping with string keys\n")),(0,o.kt)("p",null,"It's also possible to hit this error code on constraint mismatches when using tuple variadic variables as specified in ",(0,o.kt)("a",{parentName:"p",href:"https://www.python.org/dev/peps/pep-0646/"},"PEP 646"),", which are an advanced feature of pyre."),(0,o.kt)("h3",{id:"33-prohibited-any"},"33: Prohibited Any"),(0,o.kt)("p",null,"Pyre will warn on any usage of ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.Any")," when run in ",(0,o.kt)("a",{parentName:"p",href:"/docs/types-in-python#strict-mode"},"strict mode"),". ",(0,o.kt)("inlineCode",{parentName:"p"},"Any")," is an escape hatch that hides type errors and introduces potential type inconsistencies which Pyre strict is designed to make explicit. To resolve this error, replace ",(0,o.kt)("inlineCode",{parentName:"p"},"Any")," with any other annotation. Using builtins ",(0,o.kt)("inlineCode",{parentName:"p"},"object")," is acceptable if you are looking for a supertype of all classes."),(0,o.kt)("h3",{id:"34-invalid-type-variable"},"34: Invalid Type Variable"),(0,o.kt)("p",null,"Example:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class Base:\n    foo: List[T] = []\n\n$ pyre\nInvalid type variable [34]: The current class isn't generic with respect to the type variable `Variable[T]`.\n")),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def foo(x: int) -> List[T]:\n    return [x, x]\n\n$ pyre\nInvalid type variable [34]: The type variable `Variable[T]` isn't present in the function's parameters.\n")),(0,o.kt)("p",null,'Type variables can only be used as types when they have already been placed "in scope".\nA type variable can be placed into scope via:'),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},"Generic class declarations",(0,o.kt)("ul",{parentName:"li"},(0,o.kt)("li",{parentName:"ul"},"for example, ",(0,o.kt)("inlineCode",{parentName:"li"},"class C(Generic[T]):")," puts ",(0,o.kt)("inlineCode",{parentName:"li"},"T")," into scope for the body of the class"))),(0,o.kt)("li",{parentName:"ul"},"The ",(0,o.kt)("strong",{parentName:"li"},"parameter")," types of a generic function",(0,o.kt)("ul",{parentName:"li"},(0,o.kt)("li",{parentName:"ul"},"for example, ",(0,o.kt)("inlineCode",{parentName:"li"},"def foo(x: T)")," puts ",(0,o.kt)("inlineCode",{parentName:"li"},"T")," into scope for the body and return type annotation of the function")))),(0,o.kt)("p",null,"Suggested fix:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class Base(Generic[T]):\n    foo: List[T] = []\n\nbase: Base[int]\n\ndef foo(x: T) -> List[T]:\n    return [x, x]\n")),(0,o.kt)("h4",{id:"decorator-factories"},"Decorator Factories"),(0,o.kt)("p",null,"One common error is when defining a generic decorator factory. The Python type system doesn't currently place ",(0,o.kt)("inlineCode",{parentName:"p"},"T")," into scope within a ",(0,o.kt)("inlineCode",{parentName:"p"},"Callable")," type. So, it considers ",(0,o.kt)("inlineCode",{parentName:"p"},"T")," to be a type variable from the outer scope. This can lead to errors for apparently valid code:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"from typing import Callable, TypeVar\n\nT = TypeVar(\"T\")\nR = TypeVar(\"R\")\n\ndef my_decorator_factory(message: str) -> Callable[[Callable[[T], R]], Callable[[T], R]]:\n\n    def _decorator(f: Callable[[T], R]) -> Callable[[T], R]:\n\n        def _inner(x: T) -> R:\n            print(message)\n            return f(x)\n\n        return _inner\n\n    return _decorator\n\n$ pyre\nInvalid type variable [34]: The type variable `Variable[R]` isn't present in the function's parameters.\nInvalid type variable [34]: The type variable `Variable[T]` isn't present in the function's parameters.\n")),(0,o.kt)("p",null,"Suggested fix: Use a callback protocol to define the return type."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import Callable, Protocol, TypeVar\n\nT = TypeVar("T")\nR = TypeVar("R")\n\nclass MyCallableProtocol(Protocol):\n    def __call__(self, f: Callable[[T], R]) -> Callable[[T], R]: ...\n\ndef my_decorator_factory(message: str) -> MyCallableProtocol:\n\n    def _decorator(f: Callable[[T], R]) -> Callable[[T], R]:\n\n        def _inner(x: T) -> R:\n            print(message)\n            return f(x)\n\n        return _inner\n\n    return _decorator\n')),(0,o.kt)("p",null,"If you are using a ",(0,o.kt)("inlineCode",{parentName:"p"},"ParamSpec")," in your decorator, use the following:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import Awaitable, Callable, Protocol, TypeVar\nfrom pyre_extensions import ParameterSpecification\n\nR = TypeVar("R")\nP = ParameterSpecification("P")\n\nclass MyCallableProtocol(Protocol):\n    def __call__(self, f: Callable[P, Awaitable[R]]) -> Callable[P, Awaitable[R]]: ...\n\ndef my_decorator_factory(message: str) -> MyCallableProtocol:\n\n    def _decorator(f: Callable[P, Awaitable[R]]) -> Callable[P, Awaitable[R]]:\n\n        async def _inner(*args: P.args, **kwargs: P.kwargs) -> R:\n            print(message)\n            return await f(*args, **kwargs)\n\n        return _inner\n\n    return _decorator\n')),(0,o.kt)("p",null,"Note: Support for such callables is currently ",(0,o.kt)("strong",{parentName:"p"},"experimental")," and varies from one typechecker to another. This behavior may change in the future."),(0,o.kt)("h3",{id:"35-illegal-annotation-target"},"35: Illegal Annotation Target"),(0,o.kt)("p",null,"Pyre will error when a type annotation is applied to something that can't be annotated. This could happen when:"),(0,o.kt)("ol",null,(0,o.kt)("li",{parentName:"ol"},"A variable is re-annotated after first declaration or an explicity annotated function parameter is re-annotated within the function body. This is not allowed as re-annotating variables reduces readability and causes the annotation of a variable to depend on the position in control flow.")),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def transformation(p: int) -> str:\n  return str(p + 1)\n\ndef foo(x: int) -> None:\n  y: int = x + 2\n  z = x + 3\n\n  # Each of the following will produce an error\n  x: str = transformation(x)\n  y: str = transformation(y)\n  z: int = 4\n")),(0,o.kt)("p",null,"An easy fix for the first two errors is to use a new variable rather than re-annotating the old variable so it can hold a new type. For the third error, ",(0,o.kt)("inlineCode",{parentName:"p"},"z")," should have been annotated at first declaration."),(0,o.kt)("ol",{start:2},(0,o.kt)("li",{parentName:"ol"},"Trying to annotate non-self attributes, i.e annotating the attributes of a different class than the one whose scope you are in:")),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'class Foo:\n  attribute: int = 1\n\nclass Bar:\n  def __init__(self):\n    Foo.attribute: str = "hello"\n\ndef some_method() -> None:\n  Foo.attribute: int = 5\n')),(0,o.kt)("p",null,"This is not allowed as Pyre needs to be able to statically determine the type of globally accessible values, including class attributes. Even if Pyre followed control flow across functions to determine class attribute annotations, such re-annotations imply very dynamic behavior that makes the code difficult to work with."),(0,o.kt)("p",null,"The fix for this situation, similar to the case above, is to annotate the class attribute at its definition in the class that owns it and remove any annotations elsewhere. If this attribute is from a third party library, then you can add a ",(0,o.kt)("a",{parentName:"p",href:"/docs/errors#third-party-libraries"},"stub")," for the class and annotate the attribute there."),(0,o.kt)("h3",{id:"39-invalid-inheritance"},"39: Invalid Inheritance"),(0,o.kt)("p",null,"When defining a new class, Pyre will error if the base class given is not a valid parent class. This may be caused by various conditions:"),(0,o.kt)("ol",null,(0,o.kt)("li",{parentName:"ol"},"The parent class is marked as final which means it explicitly is annotated as not supporting child classes.")),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"@final\nclass Base:\n  ...\n\nclass Derived(Base): # Invalid inheritance error\n  ...\n")),(0,o.kt)("ol",{start:2},(0,o.kt)("li",{parentName:"ol"},"The expression given in the base class field is not a class at all.")),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'MY_GLOBAL: str = "string"\nclass Foo(MY_GLOBAL): # Invalid inheritance error\n  ...\n')),(0,o.kt)("p",null,"Pyre does not support dynamic expressions as base classes, even if they may evaluate to a valid class at runtime. This is because the type checker relies on building up a valid class hierarchy before it can resolve types in the Python it is analyzing. On the other hand, type aliases are equivalent to types and are acceptable as base classes."),(0,o.kt)("ol",{start:3},(0,o.kt)("li",{parentName:"ol"},"You are defining a typed dictionary that does not inherit from another typed dictionary.")),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class NonTypedDict:\n  ...\n\nclass Movie(TypedDict):\n  name: str\n  year: int\n\nclass BookBasedMovie(Movie): # No error\n  based_on: str\n\nclass BookBasedMovie(NonTypedDict): # Invalid inheritance error\n  based_on: str\n")),(0,o.kt)("p",null,"If inheriting from another typed dictionary, fields need to have a consistent type between child and parent, in order for subclassing to be sound. Similarly, a required field in the child must also be required for the parent."),(0,o.kt)("h3",{id:"40-invalid-override"},"40: Invalid Override"),(0,o.kt)("p",null,"Pyre will error when methods in a child class override those in a parent class inconsistently.\nStatic methods cannot be overwritten by non-static methods, and final methods cannot be overwritten."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class A:\n    @staticmethod\n    def foo(self) -> int:\n        pass\n\nclass B(A):\n    @classmethod # Non-static method `B.foo` cannot override a static method defined in `A`.\n    def foo(cls) -> int:\n        pass\n")),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"  from typing import final\n  class Foo:\n    @final\n    def bar(self) -> None:\n      pass\n\n  class Bar(Foo):\n    def bar(self) -> None: # Invalid override [40]: `Bar.bar` cannot override final method defined in `Foo`.\n      pass\n")),(0,o.kt)("h3",{id:"41-invalid-assignment"},"41: Invalid Assignment"),(0,o.kt)("p",null,"Pyre will error on assignments to final attributes, read-only properties, and class variables from a class instance. For example,"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class Foo:\n  field: Final[Optional[int]] = 1\n\n  def foo() -> None:\n    self.field = 2 # Invalid assignment\n\nclass Bar:\n    _x = 1\n    @property\n    def x(self) -> int:\n        return self._x\n\ndef bar(b: Bar) -> None:\n    b.x = 1 # Invalid assignment\n")),(0,o.kt)("p",null,"To fix this error, change the definition of this attribute to something that is mutable, if it is not intended to be read-only."),(0,o.kt)("h3",{id:"42-missing-overload-implementation"},"42: Missing Overload Implementation"),(0,o.kt)("p",null,"Pyre will throw this error if a source module specifies one or more overloads via ",(0,o.kt)("a",{parentName:"p",href:"https://fburl.com/d7b8cd2h"},(0,o.kt)("inlineCode",{parentName:"a"},"typing.overload"))," but fails to provide an implementation, for example:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"import typing overload\n\n@overload\ndef f(x: int) -> float:\n    ...\n\n@overload\ndef f(x: str) -> str:\n    ...\n")),(0,o.kt)("p",null,"Missing implementations are allowed in ",(0,o.kt)("inlineCode",{parentName:"p"},".pyi")," stub files."),(0,o.kt)("p",null,"To fix it, provide exactly one implementation (a function of the same name without the ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.overload")," decorator). For example above we could implement ",(0,o.kt)("inlineCode",{parentName:"p"},"f")," as follows:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"import typing overload, Union\n\n@overload\ndef f(x: int) -> float:\n    ...\n\n@overload\ndef f(x: str) -> str:\n    ...\n\n@overload\ndef f(x: str) -> str:\n    ...\n\ndef f(x: Union[int, str]) -> Union[float, str]:\n    if isinstance(x, int):\n        return float(x)\n    else:\n        return x\n")),(0,o.kt)("h3",{id:"43-incompatible-overload-implementation"},"43: Incompatible Overload Implementation"),(0,o.kt)("p",null,"Pyre will error if you define one or more overloads using ",(0,o.kt)("a",{parentName:"p",href:"https://fburl.com/d7b8cd2h"},(0,o.kt)("inlineCode",{parentName:"a"},"typing.overload")),", and your concrete implementation has an incompatible type signature."),(0,o.kt)("p",null,"For example, this code will produce an incompatible overload implementation error"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"# pyre-strict\n\nfrom typing import overload, Union\n\n@overload\ndef f(x: int) -> float:\n    ...\n\n@overload\ndef f(x: float) -> int:\n    ...\n\n@overload\ndef f(x: str) -> str:\n    ...\n\ndef f(x: Union[int, float, str]) -> Union[int, str]:\n    if isinstance(x, float):\n        return int(x)\n    elif isinstance(x, int):\n        return float(x)\n    else:\n        return x\n")),(0,o.kt)("p",null,"The problem here is that the return type ",(0,o.kt)("inlineCode",{parentName:"p"},"Union[int, str]")," is too narrow to\npermit ",(0,o.kt)("inlineCode",{parentName:"p"},"f")," to return ",(0,o.kt)("inlineCode",{parentName:"p"},"float")," when called on an ",(0,o.kt)("inlineCode",{parentName:"p"},"int")," argument."),(0,o.kt)("p",null,"You can fix this by either removing incorrect overload delcarations or making\nsure all parameters and return type annotations on the concrete\nimplementation are general enough to be consistent with the overloads:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def f(x: Union[int, float, str]) -> Union[int, float, str]:\n    <same implementation>\n")),(0,o.kt)("h3",{id:"45-invalid-class-instantiation"},"45: Invalid Class Instantiation"),(0,o.kt)("p",null,"In typed Python, certain classes are intended to represent abstract interfaces. These classes are not meant to be instantiated directly at runtime, and therefore Pyre will error on any attempt to perform such instantiations."),(0,o.kt)("p",null,"Currently, Pyre will recognize a given class ",(0,o.kt)("inlineCode",{parentName:"p"},"C")," as an interface class, if one of the following statement about ",(0,o.kt)("inlineCode",{parentName:"p"},"C")," is true:"),(0,o.kt)("ol",null,(0,o.kt)("li",{parentName:"ol"},(0,o.kt)("inlineCode",{parentName:"li"},"C")," contains one or more abstract methods that are left not overridden. Abstract methods are defined as methods that are decorated with ",(0,o.kt)("a",{parentName:"li",href:"https://docs.python.org/3/library/abc.html#abc.abstractmethod"},(0,o.kt)("inlineCode",{parentName:"a"},"@abc.abstractmethod")),"."),(0,o.kt)("li",{parentName:"ol"},(0,o.kt)("inlineCode",{parentName:"li"},"C")," directly inherits from ",(0,o.kt)("a",{parentName:"li",href:"https://docs.python.org/3/library/typing.html#typing.Protocol"},(0,o.kt)("inlineCode",{parentName:"a"},"typing.Protocol")),".")),(0,o.kt)("p",null,"To fix the error, use interface classes as type annotations only and do not instantiate any object from them. If object creation is a must, you can define subclasses of those interface classes, provide concrete implementations for all abstract methods (for case 1) or required interfaces (for case 2), and instantiate those subclasses instead."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'import abc\nfrom typing import Protocol\n\nclass Base(abc.ABC):\n    @abc.abstractmethod\n    def foo(self) -> None:\n        raise NotImplementedError\n    @abc.abstractmethod\n    def bar(self) -> str:\n        raise NotImplementedError\n\nclass Derived0(Base):\n    def foo(self) -> None:\n        print(self.bar())\n\nclass Derived1(Derived0):\n    def bar(self) -> str:\n        return "bar"\n\ndef test0() -> None:\n    base = Base()  # Error! Class `Base` contains 2 abstract methods and therefore cannot be instantiated.\n    derived0 = Derived0()  # Error! Class `Derived0` contains 1 abstract method `bar` and therefore cannot be instantiated.\n    derived1 = Derived1()  # OK\n\n\nclass MyProtocol(Protocol):\n    def baz(self, x: int) -> int:\n        ...\n\nclass MyClass:\n    def baz(self, x: int) -> int:\n        return x\n\ndef test1() -> None:\n    object0 = MyProtocol()  # Error! Class `MyProtocol` cannot be instantiated.\n    object1 = MyClass()  # OK\n')),(0,o.kt)("h3",{id:"46-invalid-type-variance"},"46: Invalid Type Variance"),(0,o.kt)("p",null,"In brief, read-only data types can be covariant, write-only data types can be contravariant, and data types that support both reads and writes must be invariant.\nIf a data type implements any functions accepting parameters of that type, we cannot guarantee that writes are not happening. If a data type implements any functions returning values of that type, we cannot guarantee that reads are not happening.\nFor example (note: int is a subclass of float in the type system and in these examples):\nWrites taking covariants:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'_T_co = typing.TypeVar("_T_co", covariant=True)\n\nclass MyList(typing.Generic[_T_co]):\n    def write(self, element: _T_co) -> None:\n        ... # adds element to list\n\ndef takes_float_list(float_list: MyList[float]) -> None:\n    float_list.write(1.0)\n\nint_list: MyList[int] = ...\ntakes_float_list(int_list)  # this call is OK because MyList is covariant: MyList[int] < MyList[float]\n# int_list contains floats\n')),(0,o.kt)("p",null,"Reads returning contravariants:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'_T_cont = typing.TypeVar("_T_cont", contravariant=True)\n\nclass MyList(typing.Generic[_T_cont]):\n    def read(self) -> _T_cont:\n        ... # returns first element from list\n\ndef takes_int_list(int_list: MyList[int]) -> int:\n   return int_list.read()\n\nfloat_list: MyList[float] = ...\ntakes_int_list(float_list)  # this call is OK because MyList is contravariant: MyList[float] < MyList[int]\n# problem with return above is clear\n')),(0,o.kt)("h3",{id:"47-invalid-method-signature"},"47: Invalid Method Signature"),(0,o.kt)("p",null,"Pyre will error if a non-static method fails to specify an expected implicit parameter like ",(0,o.kt)("inlineCode",{parentName:"p"},"self")," for an instance method or ",(0,o.kt)("inlineCode",{parentName:"p"},"cls")," for a class method, as this argument is always implicitly passed in a call and will cause a runtime crash if not specified. Additionally, Pyre will warn if this parameter is specified but typed as something incompatible with the type of the parent class."),(0,o.kt)("p",null,"Often times, the method may not need a ",(0,o.kt)("inlineCode",{parentName:"p"},"self")," or ",(0,o.kt)("inlineCode",{parentName:"p"},"cls")," and should be decorated with ",(0,o.kt)("inlineCode",{parentName:"p"},"@staticmethod")," to resolve this error. For example,"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"class Foo:\n  def foo() -> None: ...  # type error\n\nclass Foo:\n  @staticmethod\n  def foo() -> None: ... # no type error\n\nclass Foo:\n  def foo(self) -> None: ... # no type error\n")),(0,o.kt)("p",null,"Only type variables with compatible bounds can be used to annotate the ",(0,o.kt)("inlineCode",{parentName:"p"},"self")," or ",(0,o.kt)("inlineCode",{parentName:"p"},"cls")," parameter. For example,"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'P = TypeVar("T", bound="Parent")\nA = TypeVar("S", bound="ChildA")\nB = TypeVar("S", bound="ChildB")\n\nclass Parent: ...\n\nclass ChildA(Parent):\n  @classmethod\n  def foo(cls: Type[A]) -> A: ...  # no type error\n\nclass ChildB(Parent):\n  def foo(self: A) -> A: ...  # type error\n  def bar(self: B) -> B: ...  # no type error\n  def baz(self: P) -> P: ...  # no type error\n')),(0,o.kt)("h3",{id:"48-invalid-exception"},"48: Invalid Exception"),(0,o.kt)("p",null,"In python, you can only raise objects that derive from ",(0,o.kt)("inlineCode",{parentName:"p"},"BaseException")," (it's more common to subtype ",(0,o.kt)("inlineCode",{parentName:"p"},"Exception")," or one of the standard library-defined errors like ",(0,o.kt)("inlineCode",{parentName:"p"},"ValueError"),"), attempting to raise another object such as a bare string will result in a ",(0,o.kt)("inlineCode",{parentName:"p"},"TypeError"),". As a result, pyre will flag code like this:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'def f(x: int) -> None:\n    if x > 1:\n        raise "x is too big"\n')),(0,o.kt)("p",null,"To fix this, wrap the information you are trying to raise (usually an error message) in some exception type, for example:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'def f(x: int) -> None:\n    if x > 1:\n        raise ValueError("x is too big")\n')),(0,o.kt)("h3",{id:"49-unsafe-cast"},"49: Unsafe Cast"),(0,o.kt)("p",null,"Pyre supports ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.cast")," to force the type checker to accept a given type for your expression, no matter what it would otherwise infer that type to be. This is a good escape hatch but can also hide type inconsistencies and introduce unsoundness. For example:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def foo(x: int) -> str:\n    y = cast(str, x)\n    return y # No type error, even though this is unsound.\n")),(0,o.kt)("p",null,"It is safe to broaden the inferred type of a variable. In other words, casting an expression to a more general type than the type checker thinks it has is sound. If you wish to broaden the inferred type without running the risk of introducing type inconsistencies, you can use ",(0,o.kt)("inlineCode",{parentName:"p"},"pyre_extensions.safe_cast"),". This will warn if the type you are casting to is not greater than or equal to the inferred type of the expression."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def foo(x: int) -> str:\n    y = safe_cast(str, x) # Unsafe cast error\n    z = safe_cast(Union[int, str], x) # No error\n    return z # Invalid return type error\n")),(0,o.kt)("h3",{id:"51-unused-local-mode"},"51: Unused Local Mode"),(0,o.kt)("p",null,"Pyre only supports two modes of type checking, ",(0,o.kt)("a",{parentName:"p",href:"/docs/types-in-python#gradual-typing"},"unsafe")," and ",(0,o.kt)("a",{parentName:"p",href:"/docs/types-in-python#strict-mode"},"strict"),". By default, every file runs in unsafe mode, but you can change this default to strict in your ",(0,o.kt)("a",{parentName:"p",href:"/docs/configuration#configuration-files"},"configuration file"),"."),(0,o.kt)("p",null,"You can also change the type checking mode of a single file by adding a local mode in the form of a ",(0,o.kt)("inlineCode",{parentName:"p"},"# pyre-strict")," or ",(0,o.kt)("inlineCode",{parentName:"p"},"# pyre-unsafe")," comment on its own line to the file header. This will ensure that file checks under the specified mode regardless of the default."),(0,o.kt)("p",null,"If you specify more than one local mode, Pyre will error and ask you to remove all but one."),(0,o.kt)("h3",{id:"52-private-protocol-property"},"52: Private Protocol Property"),(0,o.kt)("p",null,"Python ",(0,o.kt)("a",{parentName:"p",href:"https://www.python.org/dev/peps/pep-0544/"},"Protocols"),' provide a way to statically check "duck typing", what many languages would refer to as ',(0,o.kt)("inlineCode",{parentName:"p"},"interfaces"),"."),(0,o.kt)("p",null,"Because protocols specify only an interface, they should not include ",(0,o.kt)("a",{parentName:"p",href:"https://docs.python.org/3/tutorial/classes.html#private-variables"},"private fields and methods"),", which may not be accessed outside of the class where they are defines. Pyre will complain about the following:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import Protocol\n\nclass Duck(Protocol):\n\n    def __quack(self) -> str:\n        ...\n\nclass SomeDuck:\n\n    def __quack(self) -> str:\n        return "quack"\n')),(0,o.kt)("p",null,"Usually if you are attempting to do this you have some reason not to make the protocol property public, but you can use a protected (single-leading-underscore) property instead:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import Protocol\n\nclass Duck(Protocol):\n\n    def _quack(self) -> str:\n        ...\n\nclass SomeDuck:\n\n    def _quack(self) -> str:\n        return "quack"\n')),(0,o.kt)("h3",{id:"53-missing-annotation-for-captured-variables"},"53: Missing Annotation For Captured Variables"),(0,o.kt)("p",null,"Pyre makes no attempt at trying to infer the types across function boundaries. The statement holds for nested functions as well.\nFrom a nested function's perspective, a variable defined in an nesting function behaves not too differently from a global variable. Therefore, Pyre treats such variables in the same way as it treats global variable: an explicit annotation is required if strict mode is turned on."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def outer_function0() -> int:\n    x = foo()\n    def inner_function() -> int:\n        return x  # Due to the lack of explicit annotation, Pyre will treat this variable as having type `Any`.\n    return inner_function()\n\ndef outer_function1() -> int:\n    x: int = foo()\n    def inner_function() -> int:\n        return x  # This is ok: the type of `x` is known to be `int`.\n    return inner_function()\n\ndef outer_function2() -> int:\n    x = foo()\n    def inner_function(x: int) -> int:\n        return x  # This is also ok: even though the outer `x` is not annotated, the `x` parameter of the inner function is.\n    return inner_function(x)\n")),(0,o.kt)("h3",{id:"54-invalid-typeddict-operation"},"54: Invalid TypedDict Operation"),(0,o.kt)("p",null,"In accordance with ",(0,o.kt)("a",{parentName:"p",href:"https://www.python.org/dev/peps/pep-0589/"},"PEP 598"),", code that tries to assign a value of the wrong type to a field of a ",(0,o.kt)("inlineCode",{parentName:"p"},"TypedDict")," will not typecheck:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import TypedDict\n\nclass MyDict(TypedDict):\n    value: str\n\nd: MyDict = {"value": "hello"}\nd["value"] = 5  # Invalid TypedDict operation\n')),(0,o.kt)("p",null,"To fix this you may need to change your field type to a ",(0,o.kt)("inlineCode",{parentName:"p"},"Union"),", if variable types are actually needed for a field."),(0,o.kt)("h3",{id:"55-typeddict-initialization-error"},"55: TypedDict Initialization Error"),(0,o.kt)("p",null,"Pyre will warn you when initializing a TypedDict with:"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"Missing required fields"),(0,o.kt)("pre",{parentName:"li"},(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import TypedDict\n\nclass Movie(TypedDict):\n    name: str\n    year: int\n\nmovie: Movie = {"name": "The Matrix"}\n\n$ pyre\nTypedDict initialization error [55]: Missing required field `year` for TypedDict `Movie`.\n'))),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"Incorrect field type"),(0,o.kt)("pre",{parentName:"li"},(0,o.kt)("code",{parentName:"pre",className:"language-python"},'movie: Movie = {"name": "The Matrix", "year": "1999"}\n\n$ pyre\nTypedDict initialization error [55]: Expected type `int` for `Movie` field `year` but got `str`.\n'))),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"Undefined fields"),(0,o.kt)("pre",{parentName:"li"},(0,o.kt)("code",{parentName:"pre",className:"language-python"},'movie: Movie = {"name": "The Matrix", "year": 1999, "extra_field": "hello"}\n\n$ pyre\nTypedDict initialization error [55]: TypedDict `Movie` has no field `extra_field`.\n')))),(0,o.kt)("h3",{id:"56-invalid-decoration"},"56: Invalid Decoration"),(0,o.kt)("p",null,"This error code is a catch-all for a variety of problems that can arise in the course of resolving the type of a decorated function.\nIn all of these cases, these decoration failures will lead to the function being registered with type ",(0,o.kt)("inlineCode",{parentName:"p"},"Any")," to avoid any spurious downstream errors."),(0,o.kt)("h4",{id:"pyre-was-not-able-to-infer-the-type-of-the-decorator-"},'"Pyre was not able to infer the type of the decorator ..."'),(0,o.kt)("p",null,"This should only happen when the decorator access itself is invalid, e.g. when you use a decorator which isn't declared in the stubs for a third-party library."),(0,o.kt)("h4",{id:"pyre-was-not-able-to-infer-the-type-of-argument-"},'"Pyre was not able to infer the type of argument ..."'),(0,o.kt)("p",null,'When using the "decorator factory" pattern, we need to resolve the type of both the decorator factory itself as well as the arguments passed to the decorator factory.\nThis is because the types of these arguments can alter the behavior of the returned decorator via overloads or type variables.\nHowever, this resolution has to happen early in the environment-building pipeline, when we don\'t yet have all of the context we need in order to resolve the types of arbitrary expressions.\nWe support resolving literals and simple globals as arguments, but using anything else will result in this error.'),(0,o.kt)("p",null,"To work around this, you can statically type your arguments to the decorator factory as separate globals, which can be validated later in the type-checking pipeline."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'T = TypeVar("T")\ndef decorator_factory(x: T) -> Callable[[Callable[[int], str]], Callable[[str], T]]:\n  ...\n# pyre-fixme[56]: Pyre was not able to infer the type of argument\n#  `complex_expression()` to decorator factory `decorator_factory`.\n@decorator_factory(complex_expression())\ndef foo(x: int) -> str:\n  ...\n\n\nargument: float = complex_expression()\n\n@decorator_factory(argument) # Accepted!  bar resolves to Callable[[str], float]\ndef bar(x: int) -> str:\n  ...\n')),(0,o.kt)("h4",{id:"decorator-factory-x-could-not-be-called"},'"Decorator factory ',"`","X","`",' could not be called"'),(0,o.kt)("p",null,"This corresponds to when the decorator factory access resolves to a type that is not callable (i.e. has no ",(0,o.kt)("inlineCode",{parentName:"p"},"__call__")," method)."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"not_a_factory: int = 5\n\n# pyre-fixme[56]: Decorator factory `not_a_factory` could not be called, because its\n# type `int` is not callable\n@not_a_factory(1)\ndef bar() -> None:\n  pass\n")),(0,o.kt)("h4",{id:"decorator-x-could-not-be-called"},'"Decorator ',"`","X","`",' could not be called"'),(0,o.kt)("p",null,"Similarly, these errors correspond to when the entire decorator expression (potentially including arguments to a decorator factory), resolves to a non-callable type."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def foo() -> int:\n  return 42\n\n# pyre-fixme[56]: Decorator `foo()` could not be called, because its\n# type `int` is not callable\n@foo()\ndef bar() -> None:\n  pass\n")),(0,o.kt)("h4",{id:"while-applying-decorator-factory-"},'"While applying decorator factory ..."'),(0,o.kt)("p",null,"These errors are emitted from attempting to pass the resolved factory arguments to the factory, as with any other function call."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def factory(x: str) -> Callable[[object], object]:\n  ...\n\n# pyre-fixme[56]: While applying decorator factory `factory`:\n# Expected `str` for 1st param but got `int`.\n@factory(1)\ndef foo() -> None:\n  pass\n")),(0,o.kt)("h4",{id:"while-applying-decorator-"},'"While applying decorator ..."'),(0,o.kt)("p",null,"Correspondingly, these errors are emitted from trying to pass the decorated function as an argument to the resolved decorator type."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def decorator(f: Callable[[int], str]) -> int:\n  ...\n\n# pyre-fixme[56]: While applying decorator `decorator`:\n# Expected `Callable[[int], str]` for 1st param but got `Callable[[str], int]`.\n@decorator\ndef foo(x: str) -> int:\n  return 5\n")),(0,o.kt)("h3",{id:"57-incompatible-async-generator-return-type"},"57: Incompatible Async Generator Return Type"),(0,o.kt)("p",null,"An async generator function is an ",(0,o.kt)("inlineCode",{parentName:"p"},"async")," function that contains at least one ",(0,o.kt)("inlineCode",{parentName:"p"},"yield")," statement. The Python runtime ensures that all async generator would return an async generator object. Therefore, the return type of async generator functions should always be ",(0,o.kt)("inlineCode",{parentName:"p"},"typing.AsyncGenerator")," or one of its supertypes."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"async def f() -> int:  # Error\n  yield 0\n\nfrom typing import AsyncGenerator\nasync def g() -> AsyncGenerator[int, None]:  # OK\n  if False:\n    yield 1\n")),(0,o.kt)("h3",{id:"58-unsupported-operand"},"58: Unsupported Operand"),(0,o.kt)("p",null,"Pyre will warn if an infix operator is not supported for the right or left operands provided."),(0,o.kt)("p",null,"In Python, an infix operator is converted to a method call on either of the operands - for example, ",(0,o.kt)("inlineCode",{parentName:"p"},"a < b")," is equivalent to ",(0,o.kt)("inlineCode",{parentName:"p"},"a.__lt__(b)"),". Therefore, this type error can also be considered sugar for an error that method ",(0,o.kt)("inlineCode",{parentName:"p"},"a.__lt__")," does not accept the type of ",(0,o.kt)("inlineCode",{parentName:"p"},"b")," as an argument."),(0,o.kt)("p",null,"For example,"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def foo(x: Optional[int]) -> bool:\n  return x < 0  # type error: Optional[int] is not a supported operand\n\ndef bar(x: Optional[int]) -> bool:\n  if x:\n    return x < 0  # no type error\n  return False\n")),(0,o.kt)("h3",{id:"59-duplicate-type-variables"},"59: Duplicate Type Variables"),(0,o.kt)("p",null,"This occurs when the same type variable is provided more than once to a ",(0,o.kt)("inlineCode",{parentName:"p"},"Generic")," or ",(0,o.kt)("inlineCode",{parentName:"p"},"Protocol"),". A type variable needs to be bound to a single value. Thus, if one wants two independent type variables with perhaps the same bounds or same properties, they have to be different variables."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import TypeVar, Generic\n\nT = TypeVar("T")\nS = TypeVar("S")\nclass A(Generic[T, S, T]):  # Error\n    pass\n\nclass B(Generic[T, S]):  # OK\n    pass\n')),(0,o.kt)("h3",{id:"60-unable-to-concatenate-tuple"},"60: Unable to Concatenate Tuple"),(0,o.kt)("h4",{id:"expected-to-unpack-an-iterable-"},'"Expected to unpack an iterable ..."'),(0,o.kt)("p",null,"This can occur if during concatenation of a tuple one tries to unpack a non-iterable since non-iterables can't be unpacked. Either try to unpack an iterable, or concatenate without unpacking."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"def foo(x: int, not_iterable: int, iterable: list[int]) -> None:\n  y = (x, *not_iterable)  # Error\n  z = (x, not_iterable) # OK\n  w = (x, *iterable)  # OK\n")),(0,o.kt)("h4",{id:"concatenation-not-yet-supported-for-multiple-variadic-tuples-"},'"Concatenation not yet supported for multiple variadic tuples ..."'),(0,o.kt)("p",null,"This can occur if during concatenation one tries to use multiple variadic tuples. This is due the limitations of the current type system and there is no workaround currently. One may use ",(0,o.kt)("inlineCode",{parentName:"p"},"# pyre-ignore[60]")," to suppress."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'from typing import Tuple\nfrom pyre_extensions import TypeVarTuple\n\nTs = TypeVarTuple("Ts")\n\ndef foo(xs: Tuple[*Ts]) -> None:\n  y = (*xs, *xs)  # Error\n')),(0,o.kt)("h2",{id:"suppression"},"Suppression"),(0,o.kt)("p",null,"It is not always possible to address all errors immediately \u2013 some code is too dynamic and should be refactored, other times it's ",(0,o.kt)("em",{parentName:"p"},"just not the right time")," to deal with a type error. We do encourage people to keep their type check results clean at all times and provide mechanisms to suppress errors that cannot be immediately fixed."),(0,o.kt)("h3",{id:"suppressing-individual-errors"},"Suppressing Individual Errors"),(0,o.kt)("p",null,"Pyre supports error suppression of individual errors with comments that can be placed on the line of the error or on the line preceding the error."),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("inlineCode",{parentName:"li"},"# pyre-fixme")," indicates there is an issue in the code that will be revisited later."),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("inlineCode",{parentName:"li"},"# pyre-ignore")," indicates there's an issue with the type checker or the code is too dynamic and we have decided to not fix this. If this is a Pyre bug, make sure you ",(0,o.kt)("a",{parentName:"li",href:"https://github.com/facebook/pyre/issues"},"open an issue")," on our tracker.")),(0,o.kt)("p",null,"Both comment styles allow you to suppress individual error codes as well as adding additional context."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},'def foo() -> int:\n    # pyre-fixme[7]: only suppresses return mismatches\n    return ""\n')),(0,o.kt)("p",null,"Pyre also supports ",(0,o.kt)("inlineCode",{parentName:"p"},"# type: ignore")," comments for backwards-compatibility with ",(0,o.kt)("em",{parentName:"p"},"mypy"),"."),(0,o.kt)("h3",{id:"suppressing-all-errors"},"Suppressing All Errors"),(0,o.kt)("p",null,"You can use the ",(0,o.kt)("a",{parentName:"p",href:"/docs/types-in-python#upgrade"},"Pyre upgrade tool")," to add inline error suppressions for all errors in your project."),(0,o.kt)("h3",{id:"suppressing-errors-across-files"},"Suppressing Errors Across Files"),(0,o.kt)("p",null,"You can suppress all errors in entire sections of your code by adding the path to the ",(0,o.kt)("a",{parentName:"p",href:"/docs/configuration#the-global-configuration"},(0,o.kt)("inlineCode",{parentName:"a"},"ignore_all_errors")," section of your configuration"),"."),(0,o.kt)("p",null,"Furthermore Pyre supports suppressing all errors in an individual file if you add a ",(0,o.kt)("inlineCode",{parentName:"p"},"# pyre-ignore-all-errors")," to your file. Like the other suppression comments, you can use square brackets to chose to only ignore one or more particular error types. For example, you can suppress all incompatible return type errors by adding:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-python"},"# pyre-ignore-all-errors[7]\n\ndef foo(x: int) -> str:\n  return x  # pyre will not error here\n")))}h.isMDXComponent=!0}}]);